<!DOCTYPE html><html><head><meta charset="utf-8"><style>body {
  max-width: 980px;
  border: 1px solid #ddd;
  outline: 1300px solid #fff;
  margin: 16px auto;
}

body .markdown-body
{
  padding: 45px;
}

@font-face {
  font-family: fontawesome-mini;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAzUABAAAAAAFNgAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABbAAAABwAAAAcZMzaOEdERUYAAAGIAAAAHQAAACAAOQAET1MvMgAAAagAAAA+AAAAYHqhde9jbWFwAAAB6AAAAFIAAAFa4azkLWN2dCAAAAI8AAAAKAAAACgFgwioZnBnbQAAAmQAAAGxAAACZVO0L6dnYXNwAAAEGAAAAAgAAAAIAAAAEGdseWYAAAQgAAAFDgAACMz7eroHaGVhZAAACTAAAAAwAAAANgWEOEloaGVhAAAJYAAAAB0AAAAkDGEGa2htdHgAAAmAAAAAEwAAADBEgAAQbG9jYQAACZQAAAAaAAAAGgsICJBtYXhwAAAJsAAAACAAAAAgASgBD25hbWUAAAnQAAACZwAABOD4no+3cG9zdAAADDgAAABsAAAAmF+yXM9wcmVwAAAMpAAAAC4AAAAusPIrFAAAAAEAAAAAyYlvMQAAAADLVHQgAAAAAM/u9uZ4nGNgZGBg4ANiCQYQYGJgBEJuIGYB8xgABMMAPgAAAHicY2Bm42OcwMDKwMLSw2LMwMDQBqGZihmiwHycoKCyqJjB4YPDh4NsDP+BfNb3DIuAFCOSEgUGRgAKDgt4AAB4nGNgYGBmgGAZBkYGEAgB8hjBfBYGCyDNxcDBwMTA9MHhQ9SHrA8H//9nYACyQyFs/sP86/kX8HtB9UIBIxsDXICRCUgwMaACRoZhDwA3fxKSAAAAAAHyAHABJQB/AIEAdAFGAOsBIwC/ALgAxACGAGYAugBNACcA/wCIeJxdUbtOW0EQ3Q0PA4HE2CA52hSzmZDGe6EFCcTVjWJkO4XlCGk3cpGLcQEfQIFEDdqvGaChpEibBiEXSHxCPiESM2uIojQ7O7NzzpkzS8qRqnfpa89T5ySQwt0GzTb9Tki1swD3pOvrjYy0gwdabGb0ynX7/gsGm9GUO2oA5T1vKQ8ZTTuBWrSn/tH8Cob7/B/zOxi0NNP01DoJ6SEE5ptxS4PvGc26yw/6gtXhYjAwpJim4i4/plL+tzTnasuwtZHRvIMzEfnJNEBTa20Emv7UIdXzcRRLkMumsTaYmLL+JBPBhcl0VVO1zPjawV2ys+hggyrNgQfYw1Z5DB4ODyYU0rckyiwNEfZiq8QIEZMcCjnl3Mn+pED5SBLGvElKO+OGtQbGkdfAoDZPs/88m01tbx3C+FkcwXe/GUs6+MiG2hgRYjtiKYAJREJGVfmGGs+9LAbkUvvPQJSA5fGPf50ItO7YRDyXtXUOMVYIen7b3PLLirtWuc6LQndvqmqo0inN+17OvscDnh4Lw0FjwZvP+/5Kgfo8LK40aA4EQ3o3ev+iteqIq7wXPrIn07+xWgAAAAABAAH//wAPeJyFlctvG1UUh+/12DPN1B7P3JnYjj2Ox4/MuDHxJH5N3UdaEUQLqBIkfQQioJWQ6AMEQkIqsPGCPwA1otuWSmTBhjtps2ADWbJg3EpIXbGouqSbCraJw7kzNo2dRN1cnXN1ZvT7zuuiMEI7ncizyA0URofRBJpCdbQuIFShYY+GZRrxMDVtih5TwQPHtXDFFSIKoWIbuREBjLH27Ny4MsbVx+uOJThavebgVrNRLAiYx06rXsvhxLgWx9xpfHdrs/ekc2Pl2cpPCVEITQpwbj8VQhfXSq2m+Wxqaq2D73Kne5e3NjHqQNj3CRYlJlgUl/jRNP+2Gs2pNYRQiOnmUaQDqm30KqKiTTWPWjboxnTWpvgxjXo0KrtZXAHt7hwIz0YVcj88JnKlJKi3NPAwLyDwZudSmJSMMJFDYaOkaol6XtESx3Gt1VTytdZJ3DCLeaVhVnCBH1fycHTxFXwPX+l2e3d6H/TufGGmMTLTnbSJUdo00zuBswMO/nl3YLeL/wnu9/limCuD3vC54h5NBVz6Li414AI8Vx3iiosKcQXUbrvhFFiYb++HN4DaF4XzFW0fIN4XDWJ3a3XQoq9V8WiyRmdsatV9xUcHims1JloH0YUa090G3Tro3mC6c01f+YwCPquINr1PTaCP6rVTOOmf0GE2dBc7zWIhji3/5MchSuBHgDbU99RMWt3YUNMZMJmx92YP6NsHx/5/M1yvInpnkIOM3Z8fA3JQ2lW1RFC1KaBPDFXNAHYYvGy73aYZZZ3HifbeuiVZCpwA3oQBs0wGPYJbJfg60xrKEbKiNtTe1adwrpBRwlAuQ3q3VRaX0QmQ9a49BTSCuF1MLfQ6+tinOubRBZuWPNoMevGMT+V41KitO1is3D/tpMcq1JHZqDHGs8DoYGDkxJgKjHROeTCmhZvzPm9pod+ltKm4PN7Dyvvldlpsg8D+4AUJZ3F/JBstZz7cbFRxsaAGV6yX/dkcycWf8eS3QlQea+YLjdm3yrOnrhFpUyKVvFE4lpv4bO3Svx/6F/4xmiDu/RT5iI++lko18mY1oX+5UGKR6kmVjM/Zb76yfHtxy+h/SyQ0lLdpdKy/lWB6szatetQJ8nZ80A2Qt6ift6gJeavU3BO4gtxs/KCtNPVibCtYCWY3SIlSBPKXZALXiIR9oZeJ1AuMyxLpHIy/yO7vSiSE+kZvk0ihJ30HgHfzZtEMmvV58x6dtqns0XTAW7Vdm4HJ04OCp/crOO7rd9SGxQAE/mVA9xRN+kVSMRFF6S9JFGUtthkjBA5tFCWc2l4V43Ex9GmUP3SI37Jjmir9KqlaDJ4S4JB3vuM/jzyH1+8MuoZ+QGzfnvPoJb96cZlWjMcKLfgDwB7E634JTY+asjsPzS5CiVnEWY+KsrsIN5rn3mAPjqmQBxGjcGKB9f9ZxY3mYC2L85CJ2FXIxKKyHk+dg0FHbuEc7D5NzWUX32WxFcWNGRAbvwSx0RmIXVDuYySafluQBmzA/ssqJAMLnli+WIC90Gw4lm85wcp0qjArEDPJJV/sSx4P9ungTpgMw5gVC1XO4uULq0s3v1rqLi0vX/z65vlH50f8T/RHmSPTk5xxWBWOluMT6WiOy+tdvWxlV/XQb3o3c6Ssr+r6I708GsX9/nzp1tKFh0s3v7m4vAy/Hnb/KMOvc1wump6Il48K6mGDy02X9Yd65pa+nQIjk76lWxCkG8NBCP0HQS9IpAAAeJxjYGRgYGBhcCrq214Qz2/zlUGenQEEzr/77oug/zewFbB+AHI5GJhAogBwKQ0qeJxjYGRgYH3/P46BgZ0BBNgKGBgZUAEPAE/7At0AAAB4nGNngAB2IGYjhBsYBAAIYADVAAAAAAAAAAAAAFwAyAEeAaACCgKmAx4DggRmAAAAAQAAAAwAagAEAAAAAAACAAEAAgAWAAABAAChAAAAAHiclZI7bxQxFIWPd/JkUYQChEhIyAVKgdBMskm1QkKrRETpQiLRUczueB/K7HhlOxttg8LvoKPgP9DxFxANDR0tHRWi4NjrPIBEgh1p/dm+vufcawNYFWsQmP6e4jSyQB2fI9cwj++RE9wTjyPP4LYoI89iWbyLPIe6+Bh5Hs9rryMv4GbtW+RF3EhuRa7jbrIbeQkPkjdUETOLnL0Kip4FVvAhco1RXyMnSPEz8gzWxE7kWTwUp5HnsCLeR57HW/El8gJWa58iL+JO7UfkOh4l9yMv4UnyEtvQGGECgwF66MNBooF1bGCL1ELB/TYU+ZBRlvsKQ44Se6jQ4a7hef+fh72Crv25kp+8lNWGmeKoOI5jJLb1aGIGvb6TjfWNLdkqdFvJw4l1amjlXtXRZqRN7lSRylZZyhBqpVFWmTEXgWfUrpi/hZOQXdOd4rKuXOtEWT3k5IArPRzTUU5tHKjecZkTpnVbNOnt6jzN8240GD4xtikvZW56043rPMg/dS+dlOceXoR+WPbJ55Dsekq1lJpnypsMUsYOdCW30o103Ytu/lvh+5RWFLfBjm9/N8hJntPhvx92rnoE/kyHdGasGy754kw36vsVf/lFeBi+0COu+cfgQr42G3CRpeLoZ53gmfe3X6rcKt5oVxnptHR9JS8ehVUd5wvvahN2uqxOOpMXapibI5k7Zwbt4xBSaTfoKBufhAnO/uqNcfK8OTs0OQ6l7JIqFjDhYj5WcjevCnI/1DDiI8j4ndWb/5YzDZWh79yomWXeXj7Nnw70/2TIeFPTrlSh89k1ObOSRVZWZfgF0r/zJQB4nG2JUQuCQBCEd07TTg36fb2IyBaLd3vWaUh/vmSJnvpgmG8YcmS8X3Shf3R7QA4OBUocUKHGER5NNbOOEvwc1txnuWkTRb/aPjimJ5vXabI+3VfOiyS15UWvyezM2xiGOPyuMohOH8O8JiO4Af+FsAGNAEuwCFBYsQEBjlmxRgYrWCGwEFlLsBRSWCGwgFkdsAYrXFhZsBQrAAA=) format('woff');
}

@font-face {
  font-family: octicons-anchor;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAYcAA0AAAAACjQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABMAAAABwAAAAca8vGTk9TLzIAAAFMAAAARAAAAFZG1VHVY21hcAAAAZAAAAA+AAABQgAP9AdjdnQgAAAB0AAAAAQAAAAEACICiGdhc3AAAAHUAAAACAAAAAj//wADZ2x5ZgAAAdwAAADRAAABEKyikaNoZWFkAAACsAAAAC0AAAA2AtXoA2hoZWEAAALgAAAAHAAAACQHngNFaG10eAAAAvwAAAAQAAAAEAwAACJsb2NhAAADDAAAAAoAAAAKALIAVG1heHAAAAMYAAAAHwAAACABEAB2bmFtZQAAAzgAAALBAAAFu3I9x/Nwb3N0AAAF/AAAAB0AAAAvaoFvbwAAAAEAAAAAzBdyYwAAAADP2IQvAAAAAM/bz7t4nGNgZGFgnMDAysDB1Ml0hoGBoR9CM75mMGLkYGBgYmBlZsAKAtJcUxgcPsR8iGF2+O/AEMPsznAYKMwIkgMA5REMOXicY2BgYGaAYBkGRgYQsAHyGMF8FgYFIM0ChED+h5j//yEk/3KoSgZGNgYYk4GRCUgwMaACRoZhDwCs7QgGAAAAIgKIAAAAAf//AAJ4nHWMMQrCQBBF/0zWrCCIKUQsTDCL2EXMohYGSSmorScInsRGL2DOYJe0Ntp7BK+gJ1BxF1stZvjz/v8DRghQzEc4kIgKwiAppcA9LtzKLSkdNhKFY3HF4lK69ExKslx7Xa+vPRVS43G98vG1DnkDMIBUgFN0MDXflU8tbaZOUkXUH0+U27RoRpOIyCKjbMCVejwypzJJG4jIwb43rfl6wbwanocrJm9XFYfskuVC5K/TPyczNU7b84CXcbxks1Un6H6tLH9vf2LRnn8Ax7A5WQAAAHicY2BkYGAA4teL1+yI57f5ysDNwgAC529f0kOmWRiYVgEpDgYmEA8AUzEKsQAAAHicY2BkYGB2+O/AEMPCAAJAkpEBFbAAADgKAe0EAAAiAAAAAAQAAAAEAAAAAAAAKgAqACoAiAAAeJxjYGRgYGBhsGFgYgABEMkFhAwM/xn0QAIAD6YBhwB4nI1Ty07cMBS9QwKlQapQW3VXySvEqDCZGbGaHULiIQ1FKgjWMxknMfLEke2A+IJu+wntrt/QbVf9gG75jK577Lg8K1qQPCfnnnt8fX1NRC/pmjrk/zprC+8D7tBy9DHgBXoWfQ44Av8t4Bj4Z8CLtBL9CniJluPXASf0Lm4CXqFX8Q84dOLnMB17N4c7tBo1AS/Qi+hTwBH4rwHHwN8DXqQ30XXAS7QaLwSc0Gn8NuAVWou/gFmnjLrEaEh9GmDdDGgL3B4JsrRPDU2hTOiMSuJUIdKQQayiAth69r6akSSFqIJuA19TrzCIaY8sIoxyrNIrL//pw7A2iMygkX5vDj+G+kuoLdX4GlGK/8Lnlz6/h9MpmoO9rafrz7ILXEHHaAx95s9lsI7AHNMBWEZHULnfAXwG9/ZqdzLI08iuwRloXE8kfhXYAvE23+23DU3t626rbs8/8adv+9DWknsHp3E17oCf+Z48rvEQNZ78paYM38qfk3v/u3l3u3GXN2Dmvmvpf1Srwk3pB/VSsp512bA/GG5i2WJ7wu430yQ5K3nFGiOqgtmSB5pJVSizwaacmUZzZhXLlZTq8qGGFY2YcSkqbth6aW1tRmlaCFs2016m5qn36SbJrqosG4uMV4aP2PHBmB3tjtmgN2izkGQyLWprekbIntJFing32a5rKWCN/SdSoga45EJykyQ7asZvHQ8PTm6cslIpwyeyjbVltNikc2HTR7YKh9LBl9DADC0U/jLcBZDKrMhUBfQBvXRzLtFtjU9eNHKin0x5InTqb8lNpfKv1s1xHzTXRqgKzek/mb7nB8RZTCDhGEX3kK/8Q75AmUM/eLkfA+0Hi908Kx4eNsMgudg5GLdRD7a84npi+YxNr5i5KIbW5izXas7cHXIMAau1OueZhfj+cOcP3P8MNIWLyYOBuxL6DRylJ4cAAAB4nGNgYoAALjDJyIAOWMCiTIxMLDmZedkABtIBygAAAA==) format('woff');
}

.markdown-body {
  font-family: sans-serif;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  color: #333333;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body b,
.markdown-body strong {
  font-weight: bold;
}

.markdown-body mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

.markdown-body sub,
.markdown-body sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
.markdown-body sup {
  top: -0.5em;
}
.markdown-body sub {
  bottom: -0.25em;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  -moz-box-sizing: content-box;
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre,
.markdown-body samp {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body .codehilitetable {
  border: 0;
  border-spacing: 0;
}

.markdown-body .codehilitetable tr {
  border: 0;
}

.markdown-body .codehilitetable pre,
.markdown-body .codehilitetable div.codehilite {
  margin: 0;
}

.markdown-body .linenos,
.markdown-body .code,
.markdown-body .codehilitetable td {
  border: 0;
  padding: 0;
}

.markdown-body td:not(.linenos) .linenodiv {
  padding: 0 !important;
}

.markdown-body .code {
  width: 100%;
}

.markdown-body .linenos div pre,
.markdown-body .linenodiv pre,
.markdown-body .linenodiv {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-left-radius: 3px;
  -webkit-border-bottom-left-radius: 3px;
  -moz-border-radius-topleft: 3px;
  -moz-border-radius-bottomleft: 3px;
  border-top-left-radius: 3px;
  border-bottom-left-radius: 3px;
}

.markdown-body .code div pre,
.markdown-body .code div {
  border: 0;
  -webkit-border-radius: 0;
  -moz-border-radius: 0;
  border-radius: 0;
  -webkit-border-top-right-radius: 3px;
  -webkit-border-bottom-right-radius: 3px;
  -moz-border-radius-topright: 3px;
  -moz-border-radius-bottomright: 3px;
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
}

.markdown-body * {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body input {
  font: 13px Helvetica, arial, freesans, clean, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol";
  line-height: 1.4;
}

.markdown-body a {
  color: #4183c4;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:focus,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before,
.markdown-body hr:after {
  display: table;
  content: " ";
}

.markdown-body hr:after {
  clear: both;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code,
.markdown-body pre,
.markdown-body samp {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body kbd {
  background-color: #e7e7e7;
  background-image: -moz-linear-gradient(#fefefe, #e7e7e7);
  background-image: -webkit-linear-gradient(#fefefe, #e7e7e7);
  background-image: linear-gradient(#fefefe, #e7e7e7);
  background-repeat: repeat-x;
  border-radius: 2px;
  border: 1px solid #cfcfcf;
  color: #000;
  padding: 3px 5px;
  line-height: 10px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  display: inline-block;
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body .headeranchor-link {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  display: block;
  padding-right: 6px;
  padding-left: 30px;
  margin-left: -30px;
}

.markdown-body .headeranchor-link:focus {
  outline: none;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  position: relative;
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}

.markdown-body h1 .headeranchor,
.markdown-body h2 .headeranchor,
.markdown-body h3 .headeranchor,
.markdown-body h4 .headeranchor,
.markdown-body h5 .headeranchor,
.markdown-body h6 .headeranchor {
  display: none;
  color: #000;
  vertical-align: middle;
}

.markdown-body h1:hover .headeranchor-link,
.markdown-body h2:hover .headeranchor-link,
.markdown-body h3:hover .headeranchor-link,
.markdown-body h4:hover .headeranchor-link,
.markdown-body h5:hover .headeranchor-link,
.markdown-body h6:hover .headeranchor-link {
  height: 1em;
  padding-left: 8px;
  margin-left: -30px;
  line-height: 1;
  text-decoration: none;
}

.markdown-body h1:hover .headeranchor-link .headeranchor,
.markdown-body h2:hover .headeranchor-link .headeranchor,
.markdown-body h3:hover .headeranchor-link .headeranchor,
.markdown-body h4:hover .headeranchor-link .headeranchor,
.markdown-body h5:hover .headeranchor-link .headeranchor,
.markdown-body h6:hover .headeranchor-link .headeranchor {
  display: inline-block;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre,
.markdown-body .admonition {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.markdown-body code,
.markdown-body samp {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .codehilite {
  margin-bottom: 16px;
}

.markdown-body .codehilite pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .codehilite pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

/* Admonition */
.markdown-body .admonition {
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  position: relative;
  border-radius: 3px;
  border: 1px solid #e0e0e0;
  border-left: 6px solid #333;
  padding: 10px 10px 10px 30px;
}

.markdown-body .admonition table {
  color: #333;
}

.markdown-body .admonition p {
  padding: 0;
}

.markdown-body .admonition-title {
  font-weight: bold;
  margin: 0;
}

.markdown-body .admonition>.admonition-title {
  color: #333;
}

.markdown-body .attention>.admonition-title {
  color: #a6d796;
}

.markdown-body .caution>.admonition-title {
  color: #d7a796;
}

.markdown-body .hint>.admonition-title {
  color: #96c6d7;
}

.markdown-body .danger>.admonition-title {
  color: #c25f77;
}

.markdown-body .question>.admonition-title {
  color: #96a6d7;
}

.markdown-body .note>.admonition-title {
  color: #d7c896;
}

.markdown-body .admonition:before,
.markdown-body .attention:before,
.markdown-body .caution:before,
.markdown-body .hint:before,
.markdown-body .danger:before,
.markdown-body .question:before,
.markdown-body .note:before {
  font: normal normal 16px fontawesome-mini;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  line-height: 1.5;
  color: #333;
  position: absolute;
  left: 0;
  top: 0;
  padding-top: 10px;
  padding-left: 10px;
}

.markdown-body .admonition:before {
  content: "\f056\00a0";
  color: 333;
}

.markdown-body .attention:before {
  content: "\f058\00a0";
  color: #a6d796;
}

.markdown-body .caution:before {
  content: "\f06a\00a0";
  color: #d7a796;
}

.markdown-body .hint:before {
  content: "\f05a\00a0";
  color: #96c6d7;
}

.markdown-body .danger:before {
  content: "\f057\00a0";
  color: #c25f77;
}

.markdown-body .question:before {
  content: "\f059\00a0";
  color: #96a6d7;
}

.markdown-body .note:before {
  content: "\f040\00a0";
  color: #d7c896;
}

.markdown-body .admonition::after {
  content: normal;
}

.markdown-body .attention {
  border-left: 6px solid #a6d796;
}

.markdown-body .caution {
  border-left: 6px solid #d7a796;
}

.markdown-body .hint {
  border-left: 6px solid #96c6d7;
}

.markdown-body .danger {
  border-left: 6px solid #c25f77;
}

.markdown-body .question {
  border-left: 6px solid #96a6d7;
}

.markdown-body .note {
  border-left: 6px solid #d7c896;
}

.markdown-body .admonition>*:first-child {
  margin-top: 0 !important;
}

.markdown-body .admonition>*:last-child {
  margin-bottom: 0 !important;
}

/* progress bar*/
.markdown-body .progress {
  display: block;
  width: 300px;
  margin: 10px 0;
  height: 24px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #ededed;
  position: relative;
  box-shadow: inset -1px 1px 3px rgba(0, 0, 0, .1);
}

.markdown-body .progress-label {
  position: absolute;
  text-align: center;
  font-weight: bold;
  width: 100%; margin: 0;
  line-height: 24px;
  color: #333;
  text-shadow: 1px 1px 0 #fefefe, -1px -1px 0 #fefefe, -1px 1px 0 #fefefe, 1px -1px 0 #fefefe, 0 1px 0 #fefefe, 0 -1px 0 #fefefe, 1px 0 0 #fefefe, -1px 0 0 #fefefe, 1px 1px 2px #000;
  -webkit-font-smoothing: antialiased !important;
  white-space: nowrap;
  overflow: hidden;
}

.markdown-body .progress-bar {
  height: 24px;
  float: left;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: #96c6d7;
  box-shadow: inset 0 1px 0 rgba(255, 255, 255, .5), inset 0 -1px 0 rgba(0, 0, 0, .1);
  background-size: 30px 30px;
  background-image: -webkit-linear-gradient(
    135deg, rgba(255, 255, 255, .4) 27%,
    transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%,
    transparent 77%, transparent
  );
  background-image: -moz-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -ms-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: -o-linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
  background-image: linear-gradient(
    135deg,
    rgba(255, 255, 255, .4) 27%, transparent 27%,
    transparent 52%, rgba(255, 255, 255, .4) 52%,
    rgba(255, 255, 255, .4) 77%, transparent 77%,
    transparent
  );
}

.markdown-body .progress-100plus .progress-bar {
  background-color: #a6d796;
}

.markdown-body .progress-80plus .progress-bar {
  background-color: #c6d796;
}

.markdown-body .progress-60plus .progress-bar {
  background-color: #d7c896;
}

.markdown-body .progress-40plus .progress-bar {
  background-color: #d7a796;
}

.markdown-body .progress-20plus .progress-bar {
  background-color: #d796a6;
}

.markdown-body .progress-0plus .progress-bar {
  background-color: #c25f77;
}

.markdown-body .candystripe-animate .progress-bar{
  -webkit-animation: animate-stripes 3s linear infinite;
  -moz-animation: animate-stripes 3s linear infinite;
  animation: animate-stripes 3s linear infinite;
}

@-webkit-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@-moz-keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

@keyframes animate-stripes {
  0% {
    background-position: 0 0;
  }

  100% {
    background-position: 60px 0;
  }
}

.markdown-body .gloss .progress-bar {
  box-shadow:
    inset 0 4px 12px rgba(255, 255, 255, .7),
    inset 0 -12px 0 rgba(0, 0, 0, .05);
}

/* Multimarkdown Critic Blocks */
.markdown-body .critic_mark {
  background: #ff0;
}

.markdown-body .critic_delete {
  color: #c82829;
  text-decoration: line-through;
}

.markdown-body .critic_insert {
  color: #718c00 ;
  text-decoration: underline;
}

.markdown-body .critic_comment {
  color: #8e908c;
  font-style: italic;
}

.markdown-body .headeranchor {
  font: normal normal 16px octicons-anchor;
  line-height: 1;
  display: inline-block;
  text-decoration: none;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.headeranchor:before {
  content: '\f05c';
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 4px 0.25em -20px;
  vertical-align: middle;
}

/* Media */
@media only screen and (min-width: 480px) {
  .markdown-body {
    font-size:14px;
  }
}

@media only screen and (min-width: 768px) {
  .markdown-body {
    font-size:16px;
  }
}

@media print {
  .markdown-body * {
    background: transparent !important;
    color: black !important;
    filter:none !important;
    -ms-filter: none !important;
  }

  .markdown-body {
    font-size:12pt;
    max-width:100%;
    outline:none;
    border: 0;
  }

  .markdown-body a,
  .markdown-body a:visited {
    text-decoration: underline;
  }

  .markdown-body .headeranchor-link {
    display: none;
  }

  .markdown-body a[href]:after {
    content: " (" attr(href) ")";
  }

  .markdown-body abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .markdown-body .ir a:after,
  .markdown-body a[href^="javascript:"]:after,
  .markdown-body a[href^="#"]:after {
    content: "";
  }

  .markdown-body pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  .markdown-body pre,
  .markdown-body blockquote {
    border: 1px solid #999;
    padding-right: 1em;
    page-break-inside: avoid;
  }

  .markdown-body .progress,
  .markdown-body .progress-bar {
    -moz-box-shadow: none;
    -webkit-box-shadow: none;
    box-shadow: none;
  }

  .markdown-body .progress {
    border: 1px solid #ddd;
  }

  .markdown-body .progress-bar {
    height: 22px;
    border-right: 1px solid #ddd;
  }

  .markdown-body tr,
  .markdown-body img {
    page-break-inside: avoid;
  }

  .markdown-body img {
    max-width: 100% !important;
  }

  .markdown-body p,
  .markdown-body h2,
  .markdown-body h3 {
    orphans: 3;
    widows: 3;
  }

  .markdown-body h2,
  .markdown-body h3 {
    page-break-after: avoid;
  }
}
</style><title>Expression_Matrix</title></head><body><article class="markdown-body"><h1 id="generating-a-gene-expression-matrix"><a name="user-content-generating-a-gene-expression-matrix" href="#generating-a-gene-expression-matrix" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Generating a Gene Expression Matrix</h1>
<p>Most analyses have two stages: data reduction and biological analysis.<br />
Statistical analyses of scRNA-seq data take as their starting point an <strong>expression matrix</strong>, where each row represents a gene and each column represents a sample (in scRNAseq a cell). Each entry in the matrix represents the expression level of a particular gene in a given sample (cell). In most cases the number of unique reads assigned to that gene in that sample/cell. Generating the expression matrix often involves some, or all, of the following.</p>
<div class="figure" style="text-align: center">
<img src="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/figures/flowchart2.png" alt="Flowchart of the scRNAseq analysis" width="65%" />
<p class="caption">Flowchart of scRNAseq analysis</p>
</div>

<h3 id="preprocessing-and-mapping-reads"><a name="user-content-preprocessing-and-mapping-reads" href="#preprocessing-and-mapping-reads" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Preprocessing and mapping reads</h3>
<p>Raw fastq files first need to be preprocessed, extracting any elements that are apart of the sequence read. </p>
<ul>
<li>Library Barcode (Sample Index) - Used to pool multiple samples on one sequencing lane</li>
<li>Cell Barcode (10x Barcode) – Used to identify the cell the read came from</li>
<li>Unique Molecular Index (UMI) – Used to identify reads that arise during PCR replication</li>
<li>Sequencing Reads – Used to identify the gene a read came from</li>
</ul>
<p>The remaining sequences can be mapped to a reference genome. We tend to use <a href="https://github.com/alexdobin/STAR">STAR</a> aligner. For large full-transcript datasets from well annotated organisms (e.g. mouse, human) pseudo-alignment methods (e.g. <a href="https://pachterlab.github.io/kallisto/">Kallisto</a>, <a href="http://salmon.readthedocs.io/en/latest/salmon.html">Salmon</a>) are also a good choice for alignment. For <strong>full-length</strong> datasets with tens- or hundreds of thousands of reads per cell pseudoaligners become more appealing since their run-time can be several orders of magnitude less than traditional aligners.</p>
<p><strong>Note</strong>, if <em>spike-ins</em> are used, the <em>spike-in</em> sequences should be added to the reference sequence prior to mapping.</p>
<h3 id="mapping-qc"><a name="user-content-mapping-qc" href="#mapping-qc" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Mapping QC</h3>
<p>After aligning sequence data to the genome we should evaluate the quality of the mapping. There are many ways to measure the mapping quality, including: percentage mapped, amount of reads which map to rRNA/tRNAs, proportion of uniquely mapped reads, reads mapping across splice junctions, read depth along the transcripts, etc. Methods developed for bulk RNAseq, such as <a href="http://rseqc.sourceforge.net/">RSeQC</a> and <a href="http://samtools.github.io/">samtools</a>, are applicable to single-cell data:</p>
<h3 id="gene-counting"><a name="user-content-gene-counting" href="#gene-counting" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Gene Counting</h3>
<p>STAR, Kallisto, and Salmon all quantify the expression level of each gene for<br />
each cell as a part of its output.</p>
<p>Specific steps to be performed are dependant on the type of library, the element layout of the read, and the sequencing parameters.</p>
<h2 id="scrnaseq-libraries"><a name="user-content-scrnaseq-libraries" href="#scrnaseq-libraries" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>scRNAseq Libraries</h2>
<p>Generating scRNAseq libraries is currently anactive area of research with several protocols being published in the last few years, including:</p>
<ul>
<li>CEL-seq Hashimshony, 2012</li>
<li>CEL-seq2 Hashimshony, 2016</li>
<li>Drop-seq Macosko, 2015</li>
<li>InDrop-seq Klein, 2015</li>
<li>MARS-seq Jaitin, 2014</li>
<li>SCRB-seq Soumillon, 2014</li>
<li>Seq-well Gierahn, 2017</li>
<li>Smart-seq Picelli, 2014</li>
<li>Smart-seq2 Picelli, 2014</li>
<li>SMARTer <a href="http://www.clontech.com/US/Products/cDNA_Synthesis_and_Library_Construction/Next_Gen_Sequencing_Kits/Total_RNA-Seq/Universal_RNA_Seq_Random_Primed">clontech</a></li>
<li>STRT-seq Islam, 2014</li>
</ul>
<p>Differences between the methods are are in how they capture and quantify gene expression (either <strong>full-length</strong> or <strong>tag-based</strong>). </p>
<p><strong>Full-length</strong> capture tries to achieve a uniform coverage of each transcript (many reads per transcript. <strong>Tag-based</strong> protocols only capture either the 5&rsquo;- or 3&rsquo;-end of each RNA (single read per transcript). Choice in method determines what types of analyses the data can be used for. <strong>Full-length</strong> capture can be used to distinguise different isoforms, where <strong>tag-based</strong> method can reduce the mappability and difficult to distinguish different isoforms.</p>
<ul>
<li><strong>Tag-based</strong> 3’ counting techniques<br />
– 1 read per transcript<br />
– Based on polyA <br />
– Expression analysis only<br />
– Fewer reads per cell needed (~50K reads/cell)<br />
– Less noise in expression patterns</li>
<li><strong>Full-length</strong><br />
– Based on polyA <br />
– Expression analysis<br />
– Splicing information <br />
– The more information desired beyond expression, the higher the reads needed per cell (~50K reads/cell to 10M reads/cell)</li>
</ul>
<p>For smaller experiment &lt; 5000 cells , the R packages <a href="http://bioconductor.org/packages/SingleCellExperiment"><code>SingleCellExperiment</code></a> and <a href="http://bioconductor.org/packages/scater/"><code>scater</code></a> are good choices.</p>
<h2 id="10x-genomics-generation-of-expression-matrix-with-cellranger"><a name="user-content-10x-genomics-generation-of-expression-matrix-with-cellranger" href="#10x-genomics-generation-of-expression-matrix-with-cellranger" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>10X Genomics generation of expression matrix with cellranger</h2>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger</a></p>
<p>10X Genomics cell ranger uses the <a href="https://github.com/alexdobin/STAR">STAR</a> aligner to map reads to a genome after first preprocessing them extracting sample, cell and UMI sequences.</p>
<div class="figure" style="text-align: center">
<img src="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/figures/10xread.png" alt="Elements to a 10x read" width="80%" />
<p class="caption">Elements to a 10x read</p>
</div>

<h3 id="cell-barcode-and-umi-filtering"><a name="user-content-cell-barcode-and-umi-filtering" href="#cell-barcode-and-umi-filtering" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Cell barcode and UMI filtering</h3>
<ul>
<li>
<p>Cell barcodes <br />
– Must be on static list of known cell barcode sequences <br />
– May be 1 mismatch away from the list if the mismatch occurs at a low- quality position (the barcode is then corrected).</p>
</li>
<li>
<p>UMIs (Unique Molecular Index)<br />
–  Must not be a homopolymer, e.g. AAAAAAAAAA <br />
–  Must not contain N <br />
–  Must not contain bases with base quality &lt; 10 <br />
–  UMIs that are 1 mismatch away from a higher-count UMI are corrected to that UMI if they share a cell barcode and gene. </p>
</li>
</ul>
<h3 id="marking-duplicates"><a name="user-content-marking-duplicates" href="#marking-duplicates" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Marking duplicates</h3>
<ul>
<li>Using only the confidently mapped reads with valid barcodes and UMIs, <br />
– Correct the UMIs<br />
    UMIs are corrected to more abundant UMIs that are one mismatch away in sequence. <br />
– Record which reads are duplicates of the same RNA molecule <br />
– Count only the unique UMIs as unique RNA molecules<br />
– These UMI counts form an unfiltered gene-barcode matrix. </li>
</ul>
<h3 id="filtering-cells-the-10x-way"><a name="user-content-filtering-cells-the-10x-way" href="#filtering-cells-the-10x-way" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Filtering cells (the 10x way)</h3>
<ul>
<li>Select barcodes that likely contain cells <br />
– Sum UMI counts for each barcode <br />
– Select barcodes with total UMI count &gt;10% of the 99th percentile of the expected recovered cells. ( I honestly haven&rsquo;t wrapped my head around this yet. )</li>
<li>Produces a filtered gene-barcode matrix. </li>
</ul>
<div class="figure" style="text-align: center">
<img src="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/figures/selectcells.png" alt="Filtering cells" width="65%" />
<p class="caption">Filtering cells</p>
</div>

<h3 id="marix-output"><a name="user-content-marix-output" href="#marix-output" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Marix output</h3>
<table>
<thead>
<tr>
<th align="left">Type</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Raw</td>
<td align="left">gene-barcode matrices Contains every barcode from fixed list of known-good barcode sequences. This includes background and non-cellular barcodes.</td>
</tr>
<tr>
<td align="left">Filtered</td>
<td align="left">gene-barcode matrices Contains only detected cellular barcodes.</td>
</tr>
</tbody>
</table>
<p>With 3 files needed to completely describe each gene x cell matrix</p>
<ul>
<li>matrix.mtx </li>
<li>genes.tsv</li>
<li>barcode.tsv</li>
</ul>
<h3 id="bam-output"><a name="user-content-bam-output" href="#bam-output" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Bam output</h3>
<p>10x Chromium cellular and molecular barcode information for each read is stored as TAG fields:</p>
<table>
<thead>
<tr>
<th align="left">Tag</th>
<th align="left">Type</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">CB</td>
<td align="left">Z</td>
<td align="left">Chromium cellular barcode sequence that is error-corrected and confirmed against a list of known-good barcode sequences.</td>
</tr>
<tr>
<td align="left">CR</td>
<td align="left">Z</td>
<td align="left">Chromium cellular barcode sequence as reported by the sequencer.</td>
</tr>
<tr>
<td align="left">CY</td>
<td align="left">Z</td>
<td align="left">Chromium cellular barcode read quality. Phred scores as reported by sequencer.</td>
</tr>
<tr>
<td align="left">UB</td>
<td align="left">Z</td>
<td align="left">Chromium molecular barcode sequence that is error-corrected among other molecular barcodes with the same cellular barcode and gene alignment.</td>
</tr>
<tr>
<td align="left">UR</td>
<td align="left">Z</td>
<td align="left">Chromium molecular barcode sequence as reported by the sequencer.</td>
</tr>
<tr>
<td align="left">UY</td>
<td align="left">Z</td>
<td align="left">Chromium molecular barcode read quality. Phred scores as reported by sequencer.</td>
</tr>
<tr>
<td align="left">BC</td>
<td align="left">Z</td>
<td align="left">Sample index read.</td>
</tr>
<tr>
<td align="left">QT</td>
<td align="left">Z</td>
<td align="left">Sample index read quality. Phred scores as reported by sequencer.</td>
</tr>
<tr>
<td align="left">TR</td>
<td align="left">Z</td>
<td align="left">Trimmed sequence. For the Single Cell 3&rsquo; v1 chemistry, this is trailing sequence following the UMI on Read 2. For the Single Cell 3&rsquo; v2 chemistry, this is trailing sequence following the cell and molecular barcodes on Read 1.</td>
</tr>
<tr>
<td align="left">TQ</td>
<td align="left">Z</td>
<td align="left">Trimmed sequence quality. Phred scores as reported by the sequencer.</td>
</tr>
</tbody>
</table>
<p>The following TAG fields are present if a read maps to the genome <strong>and</strong> overlaps an exon by at least one base pair. A read may align to multiple transcripts and genes, but it is only considered confidently mapped to the transcriptome it if mapped to a single gene.</p>
<table>
<thead>
<tr>
<th align="left">Tag</th>
<th align="left">Type</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">TX</td>
<td align="left">Z</td>
<td align="left">Semicolon-separated list of transcripts that are compatible with this alignment. Transcripts are specified with the transcript_id key in the reference GTF attribute column. The format of each entry is [transcript_id],[strand][pos],[cigar], where strand is either + or -, pos is the alignment offset in transcript coordinates, and cigar is the CIGAR string in transcript coordinates.</td>
</tr>
<tr>
<td align="left">GX</td>
<td align="left">Z</td>
<td align="left">Semicolon-separated list of gene IDs that are compatible with this alignment. Gene IDs are specified with the gene_id key in the reference GTF attribute column.</td>
</tr>
<tr>
<td align="left">GN</td>
<td align="left">Z</td>
<td align="left">Semicolon-separated list of gene names that are compatible with this alignment. Gene names are specified with gene_name key in the reference GTF attribute column.</td>
</tr>
<tr>
<td align="left">MM</td>
<td align="left">i</td>
<td align="left">Set to 1 if the genome-aligner (STAR) originally gave a MAPQ &lt; 255 (it multi-mapped to the genome) and Cell Ranger changed it to 255 because the read overlapped exactly one gene.</td>
</tr>
<tr>
<td align="left">RE</td>
<td align="left">A</td>
<td align="left">Single character indicating the region type of this alignment (E = exonic, N = intronic, I = intergenic).</td>
</tr>
</tbody>
</table>
<h4 id="an-example-read"><a name="user-content-an-example-read" href="#an-example-read" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>An example read</h4>
<p>J00113:284:HG27NBBXX:8:2202:16741:8594 1040 1   4491502 255 101M    *   0   0   ACTGGACAGTGATTGTGGGGAGCAAGTCCCTCAAGGCATTTAAAACAAAAATCTCGTGTAGCCCCTCAACTGTTCAAGTGGCAGACAAAATAAATTACCAT   -A-AAJJJFAFA-F&lt;&lt;&lt;JFFJFA-AF)JFFAJJFJAFFA7&lt;JFFJA&lt;JJFA&lt;F&lt;JFJJFAJAJFJFFJFJJJJJJJFJJJJJJFJFJFAFJFJJJJF&lt;&lt;&lt;A   NH:i:1  HI:i:1  AS:i:99 nM:i:0  TX:Z:ENSMUST00000027035,+2455,101M;ENSMUST00000195555,+1624,101M;ENSMUST00000192650,+3029,101M  GX:Z:ENSMUSG00000025902 GN:Z:Sox17  RE:A:E  CR:Z:CAAGATCTCGCAAACT   CY:Z:AAFFFJJJJJJJJJJJ   CB:Z:CAAGATCTCGCAAACT-1 UR:Z:GCCGAGACCT UY:Z:JJJJJJJJJJ UB:Z:GCCGAGACCT BC:Z:CAGCATCA   QT:Z:AAFFFFJJ   TR:Z:TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTAAAACTACCAGGAGGTATTTCATAGCGAGAAAACACACGCCC    TQ:Z:JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ-----------7-7&lt;))--------77&mdash;))7&lt;AA7A&ndash;))))    RG:Z:654:MissingLibrary:1:HG27NBBXX:8</p>
<h3 id="10x-genomics-sample-report"><a name="user-content-10x-genomics-sample-report" href="#10x-genomics-sample-report" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>10X genomics sample report</h3>
<p>Summary of the alignment and assignment of reads to cells and genes are present in the metrics_summary.csv.</p>
<table>
<thead>
<tr>
<th align="left">Metric</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Estimated Number of Cells</td>
<td align="left">The number of barcodes associated with cell-containing partitions, estimated from the barcode UMI count distribution.</td>
</tr>
<tr>
<td align="left">Mean Reads per Cell</td>
<td align="left">The total number of sequenced reads divided by the estimated number of cells.</td>
</tr>
<tr>
<td align="left">Median Genes per Cell</td>
<td align="left">The median number of genes detected (with nonzero UMI counts) across all cell-associated barcodes.</td>
</tr>
<tr>
<td align="left">Number of Reads</td>
<td align="left">Total number of sequenced reads.</td>
</tr>
<tr>
<td align="left">Valid Barcodes</td>
<td align="left">Fraction of reads with cell-barcodes that match the whitelist.</td>
</tr>
<tr>
<td align="left">Reads Mapped Confidently to Transcriptome</td>
<td align="left">Fraction of reads that mapped to a unique gene in the transcriptome with a high mapping quality score as reported by the aligner.</td>
</tr>
<tr>
<td align="left">Reads Mapped Confidently to Exonic Regions</td>
<td align="left">Fraction of reads that mapped to the exonic regions of the genome with a high mapping quality score as reported by the aligner.</td>
</tr>
<tr>
<td align="left">Reads Mapped Confidently to Intronic Regions</td>
<td align="left">Fraction of reads that mapped to the intronic regions of the genome with a high mapping quality score as reported by the aligner.</td>
</tr>
<tr>
<td align="left">Reads Mapped Confidently to Intergenic Regions</td>
<td align="left">Fraction of reads that mapped to the intergenic regions of the genome with a high mapping quality score as reported by the aligner.</td>
</tr>
<tr>
<td align="left">Sequencing Saturation</td>
<td align="left">The fraction of reads originating from an already-observed UMI. This is a function of library complexity and sequencing depth. More specifically, this is the fraction of confidently mapped, valid cell-barcode, valid UMI reads that had a non-unique (cell-barcode, UMI, gene). This metric was called &ldquo;cDNA PCR Duplication&rdquo; in versions of Cell Ranger prior to 1.2.</td>
</tr>
<tr>
<td align="left">Q30 Bases in Barcode</td>
<td align="left">Fraction of bases with Q-score at least 30 in the cell barcode sequences. This is the i7 index (I1) read for the Single Cell 3&rsquo; v1 chemistry and the R1 read for the Single Cell 3&rsquo; v2 chemistry.</td>
</tr>
<tr>
<td align="left">Q30 Bases in RNA Read</td>
<td align="left">Fraction of bases with Q-score at least 30 in the RNA read sequences. This is Illumina R1 for the Single Cell 3&rsquo; v1 chemistry and Illumina R2 for the Single Cell 3&rsquo; v2 chemistry.</td>
</tr>
<tr>
<td align="left">Q30 Bases in Sample Index</td>
<td align="left">Fraction of bases with Q-score at least 30 in the sample index sequences. This is the i5 index (I2) read for the Single Cell 3&rsquo; v1 chemistry and the i7 index (I1) read for the Single Cell 3&rsquo; v2 chemistry.</td>
</tr>
<tr>
<td align="left">Q30 Bases in UMI</td>
<td align="left">Fraction of bases with Q-score at least 30 in the UMI sequences. This is the R2 read for the Single Cell 3&rsquo; v1 chemistry and the R1 read for the Single Cell 3&rsquo; v2 chemistry.</td>
</tr>
<tr>
<td align="left">Fraction Reads in Cells</td>
<td align="left">The fraction of cell-barcoded, confidently mapped reads with cell-associated barcodes.</td>
</tr>
<tr>
<td align="left">Total Genes Detected</td>
<td align="left">The number of genes with at least one UMI count in any cell.</td>
</tr>
<tr>
<td align="left">Median UMI Counts per Cell</td>
<td align="left">The median number of total UMI counts across all cell-associated barcodes.</td>
</tr>
</tbody>
</table>
<h2 id="exercise"><a name="user-content-exercise" href="#exercise" class="headeranchor-link" aria-hidden="true"><span class="headeranchor"></span></a>Exercise</h2>
<ol>
<li>
<p>Log into ganesh with the username/password given</p>
</li>
<li>
<p>Link the following folder into your home directory</p>
<p>/share/biocore/workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF</p>
</li>
<li>
<p>Review the <a href="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/1-mkfastq.slurm">1-mkfastq.slurm</a> script used to generate fastq files from Illumina run output file.</p>
<ol>
<li>In the folder 2017_10X_mouse_comparative, which output folders/files were generated from this script?</li>
<li>Copy over the Reports folder and review it</li>
<li>Using zless review the first set of reads from sample 654 </li>
</ol>
</li>
<li>
<p>Review the <a href="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/2-counts.slurm">2-counts.slurm</a> script used to map fastq files.</p>
<ol>
<li>In the folder 2017_10X_mouse_comparative, which output folders/files were generated from this script?</li>
<li>Review the metrics_summary.csv file<ol>
<li>What where the total number of reads in this sample?</li>
<li>Reads Mapped Confidently to Transcriptome?</li>
<li>Sequencing Saturation?</li>
<li>Mean Reads per Cell?</li>
<li>Median UMI Counts per Cell?<br />
 A pretty version can be found in their html output <a href="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/web_summary.html">report</a>.</li>
</ol>
</li>
<li>Using samtools and rseqc evaluate the mapping file<br />
<pre><code>module load rseqc
module load samtools
bam_stat.py -i possorted_genome_bam.bam
samtools flagstat possorted_genome_bam.bam
samtools stats possorted_genome_bam.bam
</code></pre></li>
<li>head the files under raw_gene_bc_matrices and filtered_gene_bc_matrices</li>
</ol>
</li>
<li>
<p>Copy the file and extract. Review <a href="/Users/mattsettles/Workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/monday/logging-in.md">transfer</a> if needed.</p>
<p>/share/biocore/workshops/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/expression_tables.tar.gz</p>
</li>
</ol></article></body></html>